### PR TITLE
[spec/interpreter/test] Avoid lubs and glbs

### PR DESCRIPTION
**Pull request by rossberg**
This addresses #42 by:

* adding a type-annotated variant of the `select` instruction (opcode 1C, one after plain `select`),
* adding a bottom type.

Question: given the explicitly typed `select`, it would be possible to extend it to multiple values under the multi-value proposal. If we want to keep this option open then its type immediate should be a type vector, not a single type. Opinions?

### CODE REVIEW

File: document/core/binary/instructions.rst [hunk 1]
```diff
@@ -92,7 +92,7 @@ Reference Instructions
 Parametric Instructions
 ~~~~~~~~~~~~~~~~~~~~~~~
 
-:ref:`Parametric instructions <syntax-instr-parametric>` are represented by single byte codes.
+:ref:`Parametric instructions <syntax-instr-parametric>` are represented by single byte codes, possibly followd by a type annotation.
```
Comment (binji) [inline document/core/binary/instructions.rst]: sp: followed

File: document/core/syntax/instructions.rst [hunk 2]
```diff
@@ -205,12 +205,13 @@ Instructions in this group can operate on operands of any :ref:`value type <synt
...
      \dots \\&&|&
      \DROP \\&&|&
-     \SELECT
+     \SELECT~\valtype^? \\
    \end{array}
 
 The |DROP| operator simply throws away a single operand.
 
 The |SELECT| operator selects one of its first two operands based on whether its third operand is zero or not.
+It may be include by a :ref:`value type <syntax-valtype>` determining the type of these operands. If missing, the operands must be of :ref:`numeric type <syntax-numtype>`.
```
Comment (binji) [inline document/core/syntax/instructions.rst]: s/may be include by/may include/ ?

File: document/core/binary/instructions.rst (hunk 1 above)
Comment (rossberg) [reply document/core/binary/instructions.rst]: Done.

File: document/core/syntax/instructions.rst (hunk 2 above)
Comment (rossberg) [reply document/core/syntax/instructions.rst]: Done.

**Comment by rossberg** [comment]
@titzer, okay changed to enable multi-value select in the future.

**Comment by binji** [comment]
Do we want the encoding of select to be a vector of types? We could reuse the blocktype encoding instead to save 1 byte in most cases. Though we'd have to also check that the type has 0 results.

**Comment by rossberg** [comment]
@binji, good point. My reasoning for using a vector: (1) the most relevant cases where a blocktype would save a byte (numeric types) are already covered by the existing `select` instruction, which is even shorter; (2) deferring to a function type is odd here, because it would not match the actual instruction type (unlike with blocks) and only half of it is used; (3) it's going to be a rare instruction anyway, making conceptual simplicity more attractive.

**Comment by lukewagner** [comment]
I'm having a hard time imagining the use case for multi-reference-value select, much less the need to optimize this case.  Also, the need to reuse a condition multiple times seems like a reason for us to go ahead and add `dup`/`pick`, not work around their absence.

**Comment by rossberg** [comment]
@lukewagner, it's not just multi-reference, it would work for all multi-values. The use case for this is the same as for multi-values in general, e.g., imagine a select on two complex numbers. An argument against this would be that there is no native hardware support for this, though I suppose a compiler could make use of vector types in common cases?

**Comment by lukewagner** [comment]
Oh hah, sorry, not sure why I was thinking that the new select was only for ref types; commenting too early in the morning...

**Comment by binji** [comment]
@rossberg Yep, those seem like good reasons to me. Having written a few decoders now, I've found it a bit nicer when instruction+immediate has a (more or less) fixed length, but we've already ruined that with `br_table` so I suppose I don't mind adding another.

**Pull request merged**